README.md
# logstash-filter-lookup

New lookup plugin for dynamic lookup enrichment from multiple types of sources. There are a few different filters out there today that conduct different types of lookups, which may make sense being encapsulated within one lookup filter.


## Example
### 1. Web Service Example
See examples/webservice/java

### 2. Logstash config example
```
input {
	stdin{}
}
filter {
	lookup {
		field => "message"
		map_url => "http://localhost:8080/json"
		destination => "dest"
		refresh_interval => 10
	}
}
output {
	stdout{

	}
}
```
### 3. Execution example
Install with bin/plugin install logstash-filter-lookup
```sh
bin/logstash -f config/myconf.conf --log logstash.log --debug
```

### 4. Results
#### Input
200

#### Output
```
{
    :timestamp=>"2016-03-13T10:41:50.264000+0100", 
    :message=>"output received", 
    :event=>{
        "message"=>"200", 
        "@version"=>"1", 
        "@timestamp"=>"2016-03-13T09:41:49.493Z", 
        "host"=>"Mac-mini-de-angel.local", 
        "dest"=>"OK"
    }, 
    :level=>:debug, 
    :file=>"(eval)", 
    :line=>"47", 
    :method=>"output_func"
}
```

### 5. Update
bin/plugin update logstash-filter-lookup

### 6. GEM
https://rubygems.org/gems/logstash-filter-lookup

## Documentation

Logstash provides infrastructure to automatically generate documentation for this plugin. We use the asciidoc format to write documentation so any comments in the source code will be first converted into asciidoc and then into html. All plugin documentation are placed under one [central location](http://www.elastic.co/guide/en/logstash/current/).

- For formatting code or config example, you can use the asciidoc `[source,ruby]` directive
- For more asciidoc formatting tips, see the excellent reference here https://github.com/elastic/docs#asciidoc-guide

## Need Help?

Need help? Try #logstash on freenode IRC or the https://discuss.elastic.co/c/logstash discussion forum.

## Developing

### 1. Plugin Developement and Testing

#### Code
- To get started, you'll need JRuby with the Bundler gem installed.

- Create a new plugin or clone and existing from the GitHub [logstash-plugins](https://github.com/logstash-plugins) organization. We also provide [example plugins](https://github.com/logstash-plugins?query=example).

- Install dependencies
```sh
bundle install
```

#### Test

- Update your dependencies

```sh
bundle install
```

- Run tests

```sh
bundle exec rspec
```

### 2. Running your unpublished Plugin in Logstash

#### 2.1 Run in a local Logstash clone

- Edit Logstash `Gemfile` and add the local plugin path, for example:
```ruby
gem "logstash-filter-awesome", :path => "/your/local/logstash-filter-awesome"
```
- Install plugin
```sh
bin/plugin install --no-verify
```
- Run Logstash with your plugin
```sh
bin/logstash -e 'filter {awesome {}}'
```
At this point any modifications to the plugin code will be applied to this local Logstash setup. After modifying the plugin, simply rerun Logstash.

#### 2.2 Run in an installed Logstash

You can use the same **2.1** method to run your plugin in an installed Logstash by editing its `Gemfile` and pointing the `:path` to your local plugin development directory or you can build the gem and install it using:

- Build your plugin gem
```sh
gem build logstash-filter-awesome.gemspec
```
- Install the plugin from the Logstash home
```sh
bin/plugin install /your/local/plugin/logstash-filter-awesome.gem
```
- Start Logstash and proceed to test the plugin

## Contributing

All contributions are welcome: ideas, patches, documentation, bug reports, complaints, and even something you drew up on a napkin.

Programming is not a required skill. Whatever you've seen about open source and maintainers or community members  saying "send patches or die" - you will not see that here.

It is more important to the community that you are able to contribute.

For more information about contributing, see the [CONTRIBUTING](https://github.com/elastic/logstash/blob/master/CONTRIBUTING.md) file.

CHANGELOG.md
## 1.0
Initialized logstash-filter-lookup. Formerly called logstash-filter-webservicemap

## 2.0
Now can load a file. Uses RestClient

spec/filters/lookup_spec.rb
# encoding: utf-8
require "logstash/devutils/rspec/spec_helper"
require "logstash/filters/lookup"
require "webmock/rspec"
WebMock.disable_net_connect!(allow_localhost: true)

describe LogStash::Filters::LookUp do

  let(:config) { Hash.new }
  subject { described_class.new(config) }

  describe "file mapping" do
    config <<-CONFIG
      filter {
          lookup {
              fields       => ["status"]
              type => "file"
              path  => "filename"
          }
      }
    CONFIG
    content = "\
                        '200': OK\n\
                        '300': Redirect\n\
                        '400': Client Error\n\
                        '500': Server Error"
    filename = 'filename'

    RSpec.configure do |config|
      config.before(:each) do
        allow(File).to receive(:open).with(filename, 'r').and_yield( StringIO.new(content) )
      end
    end
    sample({"status" => "200"}) do
      insist { subject["lookup_result"]["status"] } == "OK"
    end
  end

  describe "webserver mapping" do
    config <<-CONFIG
      filter {
          lookup {
              fields       => ["status"]
              format => "yml"
              url  => "http://dummyurl/"
          }
      }
    CONFIG

    RSpec.configure do |config|
      config.before(:each) do
        stub_request(:any, "http://dummyurl/").
            to_return(:status => 200, :body => "\
                        '200': OK\n\
                        '300': Redirect\n\
                        '400': Client Error\n\
                        '500': Server Error", :headers => {})
      end
    end

    sample({"status" => "200"}) do
      insist { subject["lookup_result"]["status"] } == "OK"
    end
  end

  describe "webserver mapping existing YML" do
    config <<-CONFIG
      filter {
          lookup {
              fields       => ["status"]
              format => "yml"
              url  => "http://dummyurl/"
          }
      }
    CONFIG

    RSpec.configure do |config|
      config.before(:each) do
        stub_request(:get, "http://dummyurl/").
            to_return(:status => 200, :body => "\
                        '200': OK\n\
                        '300': Redirect\n\
                        '400': Client Error\n\
                        '500': Server Error", :headers => {})
      end
    end

    sample({"status" => "200"}) do
      insist { subject["lookup_result"]["status"] } == "OK"
    end
  end

  describe "webserver mapping not valid on register" do
    config <<-CONFIG
      filter {
          lookup {
              fields       => ["status"]
              format => "yml"
              url  => "http://dummyurl/"
          }
      }
    CONFIG

    RSpec.configure do |config|
      config.before(:each) do
        stub_request(:get, "http://dummyurl/").
            to_return(:status => 200, :body => "\
                        '200': OK\n\
                        '300': Redirect\n\
                        '400', Client Error\n\
                        '500': Server Error", :headers => {})
      end
    end

    sample({"status" => "200"}) do
      expect { subject }.to raise_error
    end
  end

  describe "webserver mapping JSON" do
    config <<-CONFIG
      filter {
          lookup {
              fields       => ["status"]
              format => "json"
              url  => "http://dummyurl/json"
          }
      }
    CONFIG

    RSpec.configure do |config|
      config.before(:each) do
        stub_request(:get, "http://dummyurl/json").
            to_return(:status => 200, :body => '{
	"200": "OK",
	"300": "Redirect",
	"400": "Client Error",
	"500": "Server Error"
}', :headers => {})
      end
    end

    sample({"status" => "200"}) do
      insist { subject["lookup_result"]["status"] } == "OK"
    end
  end

  describe "webserver mapping existing JSON" do
    config <<-CONFIG
      filter {
          lookup {
              fields       => ["status"]
              format => "json"
              url  => "http://dummyurl/json"
          }
      }
    CONFIG

    RSpec.configure do |config|
      config.before(:each) do
        stub_request(:get, "http://dummyurl/json").
            to_return(:status => 200, :body => '{
	"200": "OK",
	"300": "Redirect",
	"400": "Client Error",
	"500": "Server Error"
}', :headers => {})
      end
    end

    sample({"status" => "200"}) do
      insist { subject["lookup_result"]["status"] } == "OK"
    end
  end

  describe "webserver mapping not valid JSON on register" do
    config <<-CONFIG
      filter {
          lookup {
              fields       => ["status"]
              format => "json"
              url  => "http://dummyurl/json"
          }
      }
    CONFIG

    RSpec.configure do |config|
      config.before(:each) do
        stub_request(:get, "http://dummyurl/json").
            to_return(:status => 200, :body => '{
	"200": "OK",
	"300": "Redirect",
	"400": "Client Error",
	"500", "Server Error"
}', :headers => {})
      end
    end

    sample({"status" => "200"}) do
      expect { subject }.to raise_error
    end
  end

  describe "webserver mapping CSV" do
    config <<-CONFIG
      filter {
          lookup {
              fields       => ["status"]
              format => "csv"
              url  => "http://dummyurl/csv"
          }
      }
    CONFIG

    RSpec.configure do |config|
      config.before(:each) do
        stub_request(:get, "http://dummyurl/csv").
            to_return(:status => 200, :body => "200,OK
300,Redirect
400,Client Error
500,Server Error
", :headers => {})
      end
    end

    sample({"status" => "200"}) do
      insist { subject["lookup_result"]["status"] } == "OK"
    end
  end

  describe "webserver mapping existing CSV" do
    config <<-CONFIG
      filter {
          lookup {
              fields       => ["status"]
              format => "csv"
              url  => "http://dummyurl/csv"
          }
      }
    CONFIG

    RSpec.configure do |config|
      config.before(:each) do
        stub_request(:get, "http://dummyurl/csv").
            to_return(:status => 200, :body => "200,OK
300,Redirect
400,Client Error
500,Server Error
", :headers => {})
      end
    end

    sample({"status" => "200"}) do
      insist { subject["lookup_result"]["status"] } == "OK"
    end
  end

  describe "webserver mapping not valid CSV" do
    config <<-CONFIG
      filter {
          lookup {
              fields       => ["status"]
              format => "csv"
              url  => "http://dummyurl/csv"
          }
      }
    CONFIG

    RSpec.configure do |config|
      config.before(:each) do
        stub_request(:get, "http://dummyurl/csv").
            to_return(:status => 200, :body => "200OK
300,Redirect
400,Client Error
500:Server Error
", :headers => {})
      end
    end

    sample({"status" => "200"}) do
      insist { subject["lookup_result"]["status"] } == nil
    end
  end

  context "allow sprintf" do
    let(:config) do
      {
          "fields" => ["status"],
          "default_values" => {"status"=>"%{missing_mapping}"},
          "type" => "file",
          "path"=>"filename"
      }
    end
    content = "\
                        '200': OK\n\
                        '300': Redirect\n\
                        '400': Client Error\n\
                        '500': Server Error"
    filename = 'filename'
    RSpec.configure do |config|
      config.before(:each) do
        allow(File).to receive(:open).with(filename, 'r').and_yield( StringIO.new(content) )
      end
    end
    let(:event) { LogStash::Event.new("status" => "250", "missing_mapping" => "missing no match") }

    it "return the exact mapping" do
      subject.register
      subject.filter(event)
      expect(event["lookup_result"]["status"]).to eq("missing no match")
    end
  end
end


spec/filters/webservicemap_spec.rb
# encoding: utf-8
require "logstash/devutils/rspec/spec_helper"
require "logstash/filters/webservicemap"
require "webmock/rspec"
require 'digest/sha1'
WebMock.disable_net_connect!(allow_localhost: true)

describe LogStash::Filters::WebServiceMap do

  let(:config) { Hash.new }
  subject { described_class.new(config) }

  describe "webserver mapping" do
      config <<-CONFIG
      filter {
          webservicemap {
              field       => "status"
              destination => "mapping"
              map_url  => "http://dummyurl/"
          }
      }
      CONFIG

      RSpec.configure do |config|
          hash = Digest::SHA1.hexdigest 'http://dummyurl/'
          config.before(:each) do
              FileUtils.rm_rf(hash+'.yml')
              stub_request(:get, "http://dummyurl/").
              with(:headers => {'Accept'=>'*/*', 'User-Agent'=>'Ruby'}).
              to_return(:status => 200, :body => "\
                        '200': OK\n\
                        '300': Redirect\n\
                        '400': Client Error\n\
                        '500': Server Error", :headers => {})
          end
          config.after(:all) do
              FileUtils.rm_rf(hash+'.yml')
          end
      end

      sample("status" => "200") do
          insist { subject["mapping"] } == "OK"
      end
  end

  describe "webserver mapping existing YML" do
      config <<-CONFIG
      filter {
          webservicemap {
              field       => "status"
              destination => "mapping"
              map_url  => "http://dummyurl/"
          }
      }
      CONFIG

      RSpec.configure do |config|
        hash = Digest::SHA1.hexdigest 'http://dummyurl/'
          config.before(:each) do
              FileUtils.rm_rf(hash+'.yml')
              File.open(hash+'.yml', 'wb') { |f| f.write("\
                                                       '200': OKF\n\
                                                       '300': Redirect\n\
                                                       '400': Client Error\n\
                                                       '500': Server Error") }
              stub_request(:get, "http://dummyurl/").
              with(:headers => {'Accept'=>'*/*', 'User-Agent'=>'Ruby'}).
              to_return(:status => 200, :body => "\
                        '200': OK\n\
                        '300': Redirect\n\
                        '400': Client Error\n\
                        '500': Server Error", :headers => {})
          end
          config.after(:all) do
              FileUtils.rm_rf(hash+'.yml')
          end
      end

      sample("status" => "200") do
          insist { subject["mapping"] } == "OK"
      end
  end

  describe "webserver mapping not valid" do
      config <<-CONFIG
      filter {
          webservicemap {
              field       => "status"
              destination => "mapping"
              map_url  => "http://dummyurl/"
          }
      }
      CONFIG

      RSpec.configure do |config|
        hash = Digest::SHA1.hexdigest 'http://dummyurl/'
          config.before(:each) do
              FileUtils.rm_rf(hash+'.yml')
              stub_request(:get, "http://dummyurl/").
              with(:headers => {'Accept'=>'*/*', 'User-Agent'=>'Ruby'}).
              to_return(:status => 200, :body => "\
                        '200': OK\n\
                        '300': Redirect\n\
                        '400', Client Error\n\
                        '500': Server Error", :headers => {})
          end
          config.after(:all) do
              FileUtils.rm_rf(hash+'.yml')
          end
      end

      sample("status" => "200") do
          insist { subject["mapping"] } == nil
      end
  end
=begin
  describe "webserver mapping not valid existing YML" do
      config <<-CONFIG
      filter {
          webservicemap {
              field       => "status"
              destination => "mapping"
              map_url  => "http://dummyurl/"
          }
      }
      CONFIG
    context "init" do
      let(:my_map) { {"200"=>"OKF"}}
    end

      RSpec.configure do |config|

        hash = Digest::SHA1.hexdigest 'http://dummyurl/'
          config.before(:each) do
              stub_request(:get, "http://dummyurl/").
              with(:headers => {'Accept'=>'*/*', 'User-Agent'=>'Ruby'}).
              to_return(:status => 200, :body => "\
                        '200': OK\n\
                        '300': Redirect\n\
                        '400', Client Error\n\
                        '500': Server Error", :headers => {})
          end
          config.after(:all) do
              FileUtils.rm_rf(hash+'.yml')
          end
      end

      sample("status" => "200") do
          insist { subject["mapping"] } == "OKF"
      end
  end
=end
    context "allow sprintf" do
      let(:config) do
        {
          "field"       => "status",
          "destination" => "mapping",
          "fallback" => "%{missing_mapping}",
          "map_url"  => "http://dummyurl/"
        }
      end

      let(:event) { LogStash::Event.new("status" => "200", "missing_mapping" => "missing no match") }

      it "return the exact mapping" do
        subject.register
        subject.filter(event)
        expect(event["mapping"]).to eq("missing no match")
      end
    end
end


examples/webservice/java/src/main/java/es/nangel/logstash/webservicemap/test/Test.java
package es.nangel.logstash.webservicemap.test;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

@SpringBootApplication
public class Test {
    public static void main(String[] args) {
        SpringApplication.run(Test.class, args);
    }
}

lib/logstash/filters/lookup.rb
# encoding: utf-8
require 'logstash/filters/base'
require 'logstash/namespace'
require 'json'
require 'csv'
require 'rest_client'


# A general search and replace tool which uses a Web service with a YAML, CSV or JSON response to determine replacement values.
#
# The map entries can be specified with a Web service who your request produces a YML, CSV or JSON response.
#
# Operationally, if the event field specified in the `field` configuration
# matches the EXACT contents of a map entry key, the field's value will be substituted
# with the matched key's value from the map.
#
# The lookup filter will put the contents of the maching event fields into field result.

class LogStash::Filters::LookUp < LogStash::Filters::Base
  config_name "lookup"
  # The name of the logstash event field containing the value to be compared for a
  # match by the map filter (e.g. `message`, `host`, `response_code`). 
  # 
  # If this field is an array, only the first value will be used.
  config :fields, :validate => :array, :required => true

  # By default false, if false, then all the result will be stored into a field called lookup_result
  # If true, the same field used to lookup, will be replaced.
  config :override, :validate => :boolean, :default => false

  # Can be webservice or file
  config :type, :validate => :string, :default => 'webservice'

  # Path of a file
  config :path, :validate => :string

  config :format, :validate => :string, :default => 'yml'

  config :headers, :validate => :hash, :default => {  }
  config :params, :validate => :hash, :default => {  }

  #default field where store results if override option is false
  config :result_key, :validate => :string, :default => "lookup_result"

  # The full URI path of a Web service who generates an JSON, yml or CSV format response.
  # requires to append as suffix the format type. Ex: http://localhost:8080/geoPoints?type=json
  # http://localhost:8080/geoPoints/json
  # http://localhost:8080/geoPoints/csv
  # If no suffix matches, defaults to YAML
  config :url, :validate => :string

  #HTTP method
  config :method, :validate => :string, :default => "get"



  # When using a map file or url, this setting will indicate how frequently
  # (in seconds) logstash will check the YAML file or url for updates.
  config :refresh_interval, :validate => :number, :default => 300

  # In case no mapping occurs in the event (no matches), this will add a default
  # mapping string, which will always populate `field`, if the match failed.
  config :default_values, :validate => :array, :default => {}

  def get_map
    @my_map
  end

  def fill_map(data)
    get_map.merge!(data)
  end

  public
  def register
    @my_map = {}
    @next_refresh = Time.now + @refresh_interval
    load(true)
    @logger.debug? and @logger.debug("#{self.class.name}: map - ", :map => get_map)
    type = 'Exact'
    @logger.debug? and @logger.debug("#{self.class.name}: map mapping method - "+type)
  end

  # def register

  def json_loader(data)
    fill_map(JSON.parse(data))
  end

  def csv_loader(data)
    data = CSV.parse(data).inject(Hash.new) do |acc, v|
      acc[v[0]] = v[1]
      acc
    end
    fill_map(data)
  end

  def yml_loader(data)
    fill_map(YAML.load(data))
  end

  def load_data(registering, data)
    begin
      if @format.eql?('json')
        return json_loader(data)
      elsif @format.eql?('csv')
        return csv_loader(data)
      end
      yml_loader(data)
    rescue Exception => _
      if registering
        raise "#{self.class.name}: Bad Syntax in data #{data}"
      else
        @logger.warn("#{self.class.name}: Bad Syntax in map file, continuing with old map", :map_path => data)
      end
    end
  end

  def load(registering=false)
    begin
      if @type=='webservice'
        route = @url
        data = get_webservice_content(route)
      elsif @type=='file'
        route = @path
        data = get_file_content(route)
      end
      load_data(registering, data)
    rescue Exception => _
      if registering
        raise "#{self.class.name}: Failed to initialize with type #{type} and route #{route}"
      end
      @logger.warn("#{self.class.name}: Something happened with load. Continuing with old map", :type => @type, :route => route)
    end
  end

  # def load

  def get_file_content(file)
    data = ''
    File.open(file, 'r') do |read_file|
      data +=read_file.read
    end
    data
  end

  # def get_file_content

  def get_webservice_content(path)
    case @method
      when "get"
        data = RestClient.get sprint(path), sprint(@headers)
      when "post"
        data = RestClient.post sprint(path), sprint(@params), sprint(@headers)
    end
    data
  end

  # def get_webservice_content

  def sprint(hash)
    hash
  end

  public
  def filter(event)
    if @next_refresh < Time.now
      load()
      @next_refresh = Time.now + @refresh_interval
      @logger.info('downloading and refreshing map file')
    end
    begin
      matched = true
      if @override
        result = event;
      else
        event[@result_key] = {}
        result = event[@result_key]
      end
      @fields.each { |key|
        key_string = key.to_s
        if event.include?(key_string)
          val_string = event[key_string]
          if get_map.include?(val_string)
            result[key_string] = get_map[val_string]
          else
            if @default_values and @default_values.include?(key_string)
              result[key_string] = event.sprintf(@default_values[key_string])
            end
            matched = false
          end
        end
      }
      filter_matched(event) if matched and result.length == @fields.length
    rescue Exception => e
      @logger.error('Something went wrong when attempting to map from my_map', :exception => e, :field => @fields, :event => event)
    end
  end # def filter
end # class LogStash::Filters::LookUp


lib/logstash/filters/webservicemap.rb
# encoding: utf-8
require 'logstash/filters/base'
require 'logstash/namespace'
require 'open-uri'
require 'digest/sha1'
require 'json'
require 'csv'

# A general search and replace tool which uses a Web service with a YAML, CSV or JSON response to determine replacement values.
#
# The map entries can be specified with a Web service who your request produces a YML, CSV or JSON response.
#
# Operationally, if the event field specified in the `field` configuration
# matches the EXACT contents of a map entry key, the field's value will be substituted
# with the matched key's value from the map.
#
# By default, the webservicemap filter will replace the contents of the
# maching event field (in-place). However, by using the `destination`
# configuration item, you may also specify a target event field to
# populate with the new mapd value.

class LogStash::Filters::WebServiceMap < LogStash::Filters::Base
  config_name "webservicemap"

  # The name of the logstash event field containing the value to be compared for a
  # match by the map filter (e.g. `message`, `host`, `response_code`). 
  # 
  # If this field is an array, only the first value will be used.
  config :field, :validate => :string, :required => true

  # If the destination (or target) field already exists, this configuration item specifies
  # whether the filter should skip mapping (default) or overwrite the target field
  # value with the new mapping value.
  config :override, :validate => :boolean, :default => false

  # The full URI path of a Web service who generates an JSON, yml or CSV format response.
  # requires to append as suffix the format type. Ex: http://localhost:8080/geoPoints?type=json
  # http://localhost:8080/geoPoints/json
  # http://localhost:8080/geoPoints/csv
  # If no suffix matches, defaults to YAML
  config :map_url, :validate => :string, :required => true

  # When using a map file or url, this setting will indicate how frequently
  # (in seconds) logstash will check the YAML file or url for updates.
  config :refresh_interval, :validate => :number, :default => 300

  # The destination field you wish to populate with the mapd code. The default
  # is a field named `mapping`. Set this to the same value as source if you want
  # to do a substitution, in this case filter will allways succeed. This will clobber
  # the old value of the source field! 
  config :destination, :validate => :string, :default => "mapping"

  # In case no mapping occurs in the event (no matches), this will add a default
  # mapping string, which will always populate `field`, if the match failed.
  #
  # For example, if we have configured `fallback => "no match"`, using this map:
  # [source,ruby]
  #     foo: bar
  #
  # Then, if logstash received an event with the field `foo` set to `bar`, the destination
  # field would be set to `bar`. However, if logstash received an event with `foo` set to `nope`,
  # then the destination field would still be populated, but with the value of `no match`.
  # This configuration can be dynamic and include parts of the event using the `%{field}` syntax.
  config :fallback, :validate => :string

  def get_map
    @my_map
  end

  def set_map(map)
    @my_map = map;
  end

  public
  def register
    @my_map = {}
    @next_refresh = Time.now + @refresh_interval
    download_ws(@map_url, true)
    @logger.debug? and @logger.debug("#{self.class.name}: map - ", :map => get_map)
    type = 'Exact'
    @logger.debug? and @logger.debug("#{self.class.name}: map mapping method - "+type)
  end

  # def register

  def json_loader(data)
    get_map.merge!(JSON.parse(File.read(data)))
  end

  def csv_loader(data)
    data = CSV.read(data).inject(Hash.new) do |acc, v|
      acc[v[0]] = v[1]
      acc
    end
    get_map.merge!(data)
  end

  def yml_loader(data)
    get_map.merge!(YAML.load_file(data))
  end

  def load_file(registering, extension, data)
    begin
      if extension.equal?('.json')
        return json_loader(data)
      elsif extension.end_with?('.csv')
        return csv_loader(data)
      end
      yml_loader(data)
    rescue Exception => _
      if registering
        raise "#{self.class.name}: Bad Syntax in map file #{file_name}"
      else
        @logger.warn("#{self.class.name}: Bad Syntax in map file, continuing with old map", :map_path => file_name)
      end
    end
  end

  def get_extension(path)
    if path.end_with?('json')
      return '.json'
    elsif path.end_with?('csv')
      return '.csv'
    end
    '.yml'
  end

  public
  def download_ws(path, registering=false)
    extension = get_extension(path)
    temp_extension = '_temp'+extension;
    file_name = Digest::SHA1.hexdigest path
    begin
      File.open(file_name+temp_extension, 'wb') do |saved_file|
        open(path, 'rb') do |read_file|
          saved_file.write(read_file.read)
        end
      end
    rescue Exception => _
      if registering
        raise "#{self.class.name}: Failed to initialize with #{file_name} and path #{path}"
      end
      @logger.warn("#{self.class.name}: Something happened with URL. Continuing with old map", :map_path => file_name, :path => path)
    end

    begin
      load_file(registering, extension, file_name+temp_extension)
      FileUtils.mv(file_name+temp_extension, file_name+extension)
    rescue Exception => _
      FileUtils.rm_f(file_name+temp_extension)
    end
  end

  # def download_yaml

  public
  def filter(event)
    if @next_refresh < Time.now
      download_ws(@map_url)
      @next_refresh = Time.now + @refresh_interval
      @logger.info('downloading and refreshing map file')
    end

    return unless event.include?(@field) # Skip mapping in case event does not have @event field.
    return if event.include?(@destination) and not @override # Skip mapping in case @destination field already exists and @override is disabled.

    begin
      source = event[@field].is_a?(Array) ? event[@field].first.to_s : event[@field].to_s
      matched = false
      if get_map.include?(source)
        event[@destination] = get_map[source]
        matched = true
      end

      if not matched and @fallback
        event[@destination] = event.sprintf(@fallback)
        matched = true
      end
      filter_matched(event) if matched or @field == @destination
    rescue Exception => e
      @logger.error('Something went wrong when attempting to map from my_map', :exception => e, :field => @field, :event => event)
    end
  end # def filter
end # class LogStash::Filters::WebServiceMap
